### ISSUE TITLE
Grouping on the _parent field influenced by the presence of other parent-child mappings

### ISSUE THREAD

**Issue by acarstoiu** [elastic/elasticsearch#8332]
Not without surprise and after spending quite some time clueless in front of Sense, I have isolated yet another bug related to the parent-child relationship. It looks related to #5399.
Fact is that the following setup seems to work just fine (the example is simplified almost to the maximum):
```
PUT idx
{
	mappings:
	{
		Parent:
		{
			properties:
			{
				name: {type: "string", index: "not_analyzed"}
			}
		},
		Child:
		{
			_parent: { type: "Parent"},
			properties:
			{
				name: {type: "string", index: "not_analyzed"}
			}
		}
	}
}

POST idx/_bulk
{create: {_id:"A", _type: "Parent"}}
{name: "A parent"}
{create: {_parent: "A", _type: "Child"}}
{name: "child"}

GET idx/Child/_search?pretty&fields=_parent&_source=*
{
	aggregations:
	{
		per_parent:
		{
			terms: {field: "_parent"},
			aggs:
			{
				each_uid:
				{
					terms: {field: "_uid"}
				}
			}
		}
	}
}
```
The aggregation on parent field values yields a single group containing a single document of type `Child`. If another child document type is brought into play
```
PUT idx/_mapping/Deeper
{
	_parent: { type: "Child"},
	properties:
	{
		name: {type: "string", index: "not_analyzed"}
	}
}
```
things go suddenly crazy. The above aggregation indicates the `Child` document is now also a child of itself and thus appears in a second resulting group.
But wait, the evil is persistent :exclamation: Even if the last introduced type is deleted via `DELETE idx/Deeper`, the aggregation still claims the same paradox. This is even more concerning than the bug itself since it shows that permanent changes to existing inverted indices are introduced along with the third document type.

Given that there's a way to avoid this bug in _1.4.0_, when is this upcoming version due? :smirk:

**Comment by SushilShrestha** [comment]
Do we have a solution yet or are we stuck ? I am using elasticsearch 1.5.1 and stuck with same problem while aggregating by _ parent field on child type.

**Comment by clintongormley** [comment]
I believe @martijnvg is planning on removing access to the `_parent` field completely, which would make this point moot.

**Comment by jemc** [comment]
@clintongormley Note that removing access to the `_parent` field in aggregations would be a breaking change - it is clearly documented as a feature in the current guide.

See the second example on this page of the guide:
https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-parent-field.html

Also, the feature is clearly spelled out explicitly:
> The value of the `_parent` field is accessible in queries, aggregations, scripts, and when sorting

Given that it is a documented feature, I still don't understand how you can claim that its not a bug when the presence of other parent-child mappings in an otherwise well-formed aggregation as demonstrated in #14955 breaks the documented feature.

I would understand if you had decided that it was a bug that was impractical to fix, so you planned to remove the feature instead, but that doesn't seem to be what you're saying.

**Comment by jemc** [comment]
It's been suggested to me by someone else that the only way for us to be able to use both `has_child` queries and aggregate the children by parent is to use both the `_parent` meta-field and a normal field called `parent_id`, keeping them in sync by hand with every operations.

However, this seems fragile and honestly rather pathetic when the `_parent` meta-field is documented as being "accessible in queries, aggregations, scripts, and when sorting".

**Comment by acarstoiu** [comment]
I couldn't agree more, but I suppose @clintongormley already knows it.

**Comment by clintongormley** [comment]
The `_parent` field is no longer available for aggregations in 5.0.  Closing
